settings: Update save and discard buttons to redesigned button styles.

sayamsamal opened the pull request:
This PR updates the save and discard buttons in the setting modals to use redesigned button styles along with the new loading indicator.

Fixes: [#issues > 🎯 save/discard buttons misaligned on some browsers @ 💬](https://chat.zulip.org/#narrow/channel/9-issues/topic/.F0.9F.8E.AF.20save.2Fdiscard.20buttons.20misaligned.20on.20some.20browsers/near/2124637)

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
| Before | After |
|--------|--------|
| ![Screenshot 2025-03-20 at 2 55 42 AM](https://github.com/user-attachments/assets/bc0ba408-32b5-4c17-afc5-f7c4bd047a1c) | ![Screenshot 2025-03-20 at 2 54 44 AM](https://github.com/user-attachments/assets/77393ef1-517e-4607-925a-d7543bfb641c) |
| ![Screenshot 2025-03-20 at 2 56 15 AM](https://github.com/user-attachments/assets/5658a59b-d256-4ad7-a234-3544d373f5b2) | ![Screenshot 2025-03-20 at 12 19 05 AM](https://github.com/user-attachments/assets/300d3a58-0a59-49de-b7d3-4a4b95cc8934) |
| ![Screenshot 2025-03-20 at 2 57 11 AM](https://github.com/user-attachments/assets/3a7189fd-14f5-4a6e-beb1-bdcf0cbf994b) | ![Screenshot 2025-03-20 at 12 18 25 AM](https://github.com/user-attachments/assets/9bebf637-75e4-414e-ba47-be1b6ede7eac) | 

**Responsiveness and loading indicator**
| Before | After |
|--------|--------|
| ![save_discard_button_resp_before](https://github.com/user-attachments/assets/6846a176-2b13-450e-aeb7-6a0d00bdc918) | ![save_discard_button_resp_after](https://github.com/user-attachments/assets/cb60eee0-04f3-432f-8b7b-40c26590c759) | 

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented:
This seems plausible, but is failing node tests that still reference the only syntax, and I'd like @sahil839 to carefully test this.

sahil839 commented:
I tested changing and discarding settings in realm, streams and group settings UI on various font-sizes as well and it looks good overall. Also, tested disabled state of buttons which also work well.

timabbott commented:
I'll plan to include this in 10.1.

sayamsamal commented:
![save_discard_responsive_flexbox](https://github.com/user-attachments/assets/9e148e65-bdc8-4a00-b933-0296e4ac7432)

Since the breakpoint was removed to allow for a more flexible UI, the "Save changes" button moves on the top when it is clicked as we hide the "Discard" button, do we proceed as such? Or should we be not hiding the discard button and instead just disable it, so that the buttons stay at the next line at smaller viewports?

sayamsamal commented:
@sahil839 could you please verify if the changes look good?

sahil839 commented:
There is some change in saving indicator like the one in "Preferences" panel, loading indicator looks misaligned. Though, the design was not perfect before these changes as well, so maybe we can fix that completely.

Looks good otherwise in testing. Not sure about the commit structure with respect to CSS changes, as some changes done in first commit are changed in second commit. So, maybe that can be fixed.

alya commented:
You can ping me for a review once @sahil839 's feedback has been addressed.

sayamsamal commented:
## Updates
- Fixed the commit structure issue as noted by @sahil839.
> Looks good otherwise in testing. Not sure about the commit structure with respect to CSS changes, as some changes done in first commit are changed in second commit. So, maybe that can be fixed.
- Fixed the saving indicator styling

| Before | After |
|--------|--------|
| ![saving_indicator-before](https://github.com/user-attachments/assets/2c1696c3-bead-43a8-be25-b52363d0640b) | ![saving_indicator-after](https://github.com/user-attachments/assets/d2b45f2f-d1c8-4bb6-a9b9-93cc98b68dfa) |
| ![preferences_settings-before](https://github.com/user-attachments/assets/fe4abda2-d972-499e-8711-a5f80b4d6995) | ![preferences_settings-after](https://github.com/user-attachments/assets/ae7007e0-70de-40de-be00-e5b189de1694) |
| ![privacy_settings-before](https://github.com/user-attachments/assets/6d0d9c1d-69dc-4162-b51b-19af9ef1835a) | ![privacy_settings-after](https://github.com/user-attachments/assets/c0d9eaeb-5cb6-4351-9a57-0b5153fa6789) |
| ![topic_settings-before](https://github.com/user-attachments/assets/733ae3c2-0175-4deb-958e-0dc48cb768d3) | ![topic_settings-after](https://github.com/user-attachments/assets/e779545f-5f11-4d35-9a6f-0e4e5cd32245) |

sayamsamal commented:
@alya pinging you for a review as I have addressed @sahil839's feedback.

timabbott commented:
I read the code, and it looks plausible. I noticed one anomaly while testing:
- If you set, in the browser inspector, your network to my "On Mars" configuration (i.e., infinite latency), and try to save a change, the loading spinner gets hidden/disappears even though we've not received a response from the server.

I confirmed that happens on `main`, so it's a bug we should fix (and backport the fix for) separately. 

Probably worth @sahil839 doing another review pass just because it is making changes to the generic `subsection-header` pass, and that is fairly risky.

sahil839 commented on web/tests/settings_org.test.cjs:
```diff
@@ -61,14 +65,14 @@ function createSaveButtons(subsection) {
     const $save_button_controls = $(".save-button-controls");
     const $stub_save_button = $(".save-discard-widget-button.save-button");
```
We do not use `save-discard-widget-button` anymore, so maybe we can update them here as well.

sayamsamal replied:
Updated the selectors to just `.save-button` and `.discard-button`.

sahil839 commented on web/templates/settings/account_settings.hbs:
```diff
@@ -91,8 +91,11 @@
+            <div class="subsection-header">
+                <h3 class="inline-block">{{t "Privacy" }}</h3>
+                <div class="alert-notification privacy-setting-status"></div>
+
```
This extra line can be removed.

sayamsamal replied:
Oops, my bad, I overlooked this. FIxed!

sahil839 reviewed (commented):
Posted a couple of comments.
- Space between heading and indicator is more here than at other places.
![image](https://github.com/user-attachments/assets/87d67d27-2822-4246-8bfb-a013b9af7759)
- In group permissions panel, save discard buttons render on next line when there is enough space
![image](https://github.com/user-attachments/assets/68884e4c-a969-4135-ac9d-0eb66f5ab2f9)

sayamsamal commented:
Here are the "Saved" button with the "Primary" intent in the 3 attention levels:
![Screenshot 2025-04-01 at 10 43 23 PM](https://github.com/user-attachments/assets/a54729a0-7f2b-45ab-a022-5c01932e6f52)
![Screenshot 2025-04-01 at 10 43 39 PM](https://github.com/user-attachments/assets/bebc114e-c6d3-42d2-90be-f80dd8873cbd)
![Screenshot 2025-04-01 at 10 43 50 PM](https://github.com/user-attachments/assets/8dbb068b-eb5e-4874-8210-f47a9f97fe48)

alya commented:
Low emphasis looks good to me.

sayamsamal commented:
Updated the "Saved" state of the save button to show up as **borderless attention + success intent** action button.

| Before | After |
|--------|--------|
| ![Screenshot 2025-04-02 at 4 42 05 AM](https://github.com/user-attachments/assets/3a15de50-84e3-4069-91ad-4f633d1799d1) | ![Screenshot 2025-04-02 at 4 40 52 AM](https://github.com/user-attachments/assets/53bd4b08-0817-4658-a7b3-22187fde291c) |
| ![Screenshot 2025-04-02 at 4 41 58 AM](https://github.com/user-attachments/assets/d041af26-12f1-45b6-9297-1c4d3095081a) | ![Screenshot 2025-04-02 at 4 41 24 AM](https://github.com/user-attachments/assets/f298f57a-7a39-4357-9412-711c6ba70b83) |

alya commented:
I feel like I'd like the "Saved" indicator to be shown for another second or two, as it's more subtle.

I also think we don't want a hover color change for it, and it should use a regular mouse pointer. Otherwise, we're misleading users into thinking that clicking it will do something.

alya commented:
Is this PR ready for integration review at this point?

sayamsamal commented:
> I have no idea whether this is a regression, but when I make my connection 3G, sometimes after the loading indicator I see "Saved", and sometimes it's "Save changes". I was playing with the "Channel permissions" section in org permissions (seeing "Save changes" just happened for DM permissions). I feel like it must be a regression, as it's happening a lot?

It looks like this has to be fixed for the improvements for the "Saved" button, such as giving it time to disappear. I'm working on a solution, and will update the PR very soon.

sayamsamal commented:
Updated the PR to improve the saving experience of the save discard widget (screencast shows the action at 3G throttling):

![save_discard_widget_working_now](https://github.com/user-attachments/assets/d51b23bf-b891-4c71-9159-22f501c55af6)

sayamsamal commented:
@sahil839 I have added a commit which involves the sync realm settings codepath. Could you have look to see if I'm not messing anything else in the process?

sayamsamal commented:
> I read the code, and it looks plausible. I noticed one anomaly while testing:
> 
> * If you set, in the browser inspector, your network to my "On Mars" configuration (i.e., infinite latency), and try to save a change, the loading spinner gets hidden/disappears even though we've not received a response from the server.
> 
> I confirmed that happens on `main`, so it's a bug we should fix (and backport the fix for) separately.

Also @timabbott could you check if this is fixed as well? I think this was due to the server event call making its changes before the success callback could do it's thing.

timabbott commented:
Backported to 10.x in #34576.
